lib/1.4/dml-builtins.dml

```diff
@@ -578,6 +594,191 @@ template device {
+      models.
+*/
+template _subdevice_thread_aware is subdevice {
+    param _each_subdevice_thread_aware : sequence(_subdevice_thread_aware);
+    param _each_subdevice_thread_aware =
+        each _subdevice_thread_aware in (this);
+
+    shared method _concurrency_group_attr_list_len() -> (uint32) {
+        local uint32 len = 1;
+        foreach sd in (_each_subdevice_thread_aware) {
+            len += sd._concurrency_group_attr_list_len();
+        }
+        return len;
+    }
+
+    shared method _populate_concurrency_group_attr_list(attr_value_t *vals)
+        -> (uint64) {
+        local uint64 i = 0;
+        vals[i++] = SIM_make_attr_object(this._port_obj());
+        foreach sd in (_each_subdevice_thread_aware) {
+            i += sd._populate_concurrency_group_attr_list(vals + i);
+        }
+        return i;
+    }
+}
+
+template _thread_aware is device {
+    param __thread_aware = true;
+    param init_concurrency_mode default Sim_Concurrency_Mode_Serialized;
+    param supported_modes default Sim_Concurrency_Mode_Serialized
+                                  | Sim_Concurrency_Mode_Serialized_Memory;
+
+    group _init_current_concurrency_mode is init {
+        method init() {
+            dev.concurrency_mode.current = dev.init_concurrency_mode;
+            assert (dev.concurrency_mode.current
+                    & dev.concurrency_mode.current - 1) == 0;
+            assert (dev.concurrency_mode.current
+                    & dev.supported_modes) != 0;
+        }
+    }
+
+    implement concurrency_mode {
+        saved concurrency_mode_t current;
+
+        method supported_modes() -> (concurrency_mode_t) {
+            return dev.supported_modes;
+        }
+
+        method switch_mode(concurrency_mode_t mode) {
+            assert (mode & mode - 1) == 0;
+            assert (mode & dev.supported_modes) != 0;
+            current = mode;
+        }
+
+        method current_mode() -> (concurrency_mode_t) {
+            return current;
+        }
+    }
+
+    in each (subdevice) { is _subdevice_thread_aware; }
+
+    implement concurrency_group {
+        method execution_group(uint32 group_index) -> (attr_value_t) {
+            if (group_index > 0) return SIM_make_attr_nil();
+
+            local uint32 len = 1
+                             + (each bank in (dev)).len
+                             + (each port in (dev)).len;
+            foreach sd in (each _subdevice_thread_aware in (dev)) {
+                len += sd._concurrency_group_attr_list_len();
+            }
+
+            local attr_value_t list = SIM_alloc_attr_list(len);
+            local attr_value_t *vals = SIM_attr_list(list);
+
+            local uint64 i = 0;
+            vals[i++] = SIM_make_attr_object(dev.obj);
+            foreach x in (each bank in (dev)) {
+                vals[i++] = SIM_make_attr_object(x._bank_obj());
+            }
+            foreach x in (each port in (dev)) {
```

Comment:
Absolutely not! All port objects of a DML device are modelled by the same DML model as that which models the device (with one notable exception), and may freely interact directly with the device's state like any other part of the model! They have to be part of the same thread domain!

If you mean "should include the device only because the massive list will affect performance" then see my previous comment.
Note that the API refmanual explicitly states that all ports of a device will be placed in the same thread domain as its parent unless that port either implements the `concurrency_mode` interface or is explicitly grouped with an object of a different thread domain. Because of that, this `concurrency_group` implementation is redundant. The reason I'm implementing it is because that rule is problematic, and will likely be changed. This has to do with that notable exception I mentioned earlier -- the `init_as_subobj` connect template registers a port object for the subobject, but unlike all other ports that object has no relation to the DML model. In fact, it's in the general case unsafe to place that subobject in the same thread domain as the thread-aware DML device. That's why I added code to `init_as_subobj` template so that it can't be used if the device model is thread-aware.

Reply:
>All port objects of a DML device are modelled by the same DML model as that which models the device (with one notable exception), and may freely interact directly with the device's state like any other part of the model! They have to be part of the same thread domain!
 
 In this case it's a one unit and should be added as a whole device.

Reply:
WDYM by a "whole device"? It's a device object with port subobjects; that's how Simics works. You can't have it be just one object.

Reply:
Discussed in a call: the chief -- and valid -- concern is that the user should have the ability to add other objects to the device's thread domain; i.e. the user should have control over the `concurrency_group` interface implementation. This means at minimum the methods need to be overridable, but really, given how the current implementation is just a haphazard means of trying to be future-compatible with possible changes to the thread domain API, it's better to remove the `concurrency_group` implementation entirely. It is entirely redundant with the rules as they are right now -- as questionable as they might be -- and trying to predict how they may change is just pointless.

Reply:
It's not possible to split the DML model to be in the different concurrency groups. The DML model object (whole device) should be included into the group in this case.

Reply:
> It's not possible to split the DML model to be in the different concurrency groups
_Technically_ incorrect, but for all intents and purposes; yes, that's true with the thread domain grouping rules being as they are. The original idea behind the `concurrency_group` implementation was to somehow be future-proof with if the thread domain grouping rules were to change (as the desire to have `init_as_subobj` work make them flawed.) Regardless, this future-proofing attempt very much premature.